src/coprocessor/dag/expr/compare.rs

```diff
@@ -161,6 +164,115 @@ impl FnCall {
+    }
+
+    pub fn greatest_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), max)
+    }
+
+    pub fn greatest_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut greatest = Time::new(
+            ctx.cfg.tz.timestamp(mysql::time::ZERO_TIMESTAMP, 0),
+            mysql::types::DATETIME,
+            mysql::DEFAULT_FSP,
+        )?;
+
+        for exp in &self.children {
+            let val = try_opt!(exp.eval_string(ctx, row));
+            let s = str::from_utf8(&val)?;
+            match Time::parse_datetime(s, self.tp.get_decimal() as i8, &ctx.cfg.tz) {
+                Ok(t) => greatest = max(greatest, t),
+                Err(_) => {
+                    if let Err(e) = ctx.handle_invalid_time_error(Error::invalid_time_format(s)) {
+                        return Err(e);
+                    }
+                }
+            }
+        }
+
+        Ok(Some(Cow::Owned(greatest.to_string().into_bytes())))
+    }
+
+    pub fn greatest_string<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        do_get_extremum(self, |e| e.eval_string(ctx, row), max)
+    }
+
+    pub fn least_int(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<i64>> {
+        do_get_extremum(self, |e| e.eval_int(ctx, row), min)
+    }
+
+    pub fn least_real(&self, ctx: &mut EvalContext, row: &[Datum]) -> Result<Option<f64>> {
+        do_get_extremum(self, |e| e.eval_real(ctx, row), |x, y| x.min(y))
+    }
+
+    pub fn least_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        do_get_extremum(self, |e| e.eval_decimal(ctx, row), min)
+    }
+
+    pub fn least_time<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let mut res = Cow::from(vec![]);
+        let mut least = Time::new(
+            ctx.cfg.tz.timestamp(mysql::time::MAX_TIMESTAMP, 0),
+            mysql::types::DATETIME,
+            mysql::DEFAULT_FSP,
+        )?;
+        let mut find_invalid_time = false;
+
+        for exp in &self.children {
+            let val = try_opt!(exp.eval_string(ctx, row));
+            let s = str::from_utf8(&val)?;
+            match Time::parse_datetime(s, self.tp.get_decimal() as i8, &ctx.cfg.tz) {
+                Ok(t) => least = min(least, t),
+                Err(_) => match ctx.handle_invalid_time_error(Error::invalid_time_format(s)) {
+                    Err(e) => return Err(e),
+                    _ => if !find_invalid_time {
+                        res = val.clone();
+                        find_invalid_time = true;
```

Comment:
Seems `find_invalid_time ` would never be set to false again once it's true, so why not return directly?

Reply:
Because we still need to check rest of the child expressions and handle NULL/error cases(the 'try_opt!' macro)